### ISSUE TITLE
ASAN errors when using spdk-cli to delete bdev

### ISSUE THREAD

**Issue by ChenWang01** [spdk/spdk#369]
The SPDK target will report ASAN AddressSanitizer error when using spdk-cli to delete bdev

## Expected Behavior
bdev should be deleted.

## Current Behavior
target:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==2885==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000563895 bp 0x000000074740 sp 0x7ffd58f6bb10 T0)
==2885==The signal is caused by a WRITE memory access.
==2885==Hint: address points to the zero page.
    #0 0x563894 in malloc_elem_free (/home/wangchen/spdk/app/spdk_tgt/spdk_tgt+0x563894)
    #1 0x562cae in rte_free (/home/wangchen/spdk/app/spdk_tgt/spdk_tgt+0x562cae)
    #2 0x42195f in _spdk_bs_dev_destroy /home/wangchen/spdk/lib/blob/blobstore.c:2266
    #3 0x51b8d0 in _finish_unregister /home/wangchen/spdk/lib/thread/thread.c:361
    #4 0x516265 in _spdk_event_queue_run_batch /home/wangchen/spdk/lib/event/reactor.c:207
    #5 0x516265 in _spdk_reactor_run /home/wangchen/spdk/lib/event/reactor.c:506
    #6 0x5172a2 in spdk_reactors_start /home/wangchen/spdk/lib/event/reactor.c:692
    #7 0x513ea6 in spdk_app_start /home/wangchen/spdk/lib/event/app.c:576
    #8 0x411bb1 in main /home/wangchen/spdk/app/spdk_tgt/spdk_tgt.c:121
    #9 0x7fcb83d7718a in __libc_start_main (/lib64/libc.so.6+0x2318a)
    #10 0x414bd9 in _start (/home/wangchen/spdk/app/spdk_tgt/spdk_tgt+0x414bd9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/home/wangchen/spdk/app/spdk_tgt/spdk_tgt+0x563894) in malloc_elem_free
==2885==ABORTING
```
spdk-cli:
```
/bdevs/nvme> delete nvme0n1
Connection closed with partial response:

[Errno 32] Broken pipe

```
## Steps to Reproduce
```
1. ./app/spdk_tgt/spdk_tgt
2. ./scripts/spdkcli.py
3. cd bdevs/nvme
4. create name=nvme0 trtype=PCIe traddr=0000.03.00.0
5. create name=nvme0 trtype=PCIe traddr=0000.04.00.0
6. delete nvme0n1
```
## Context (Environment including OS version, SPDK version, etc.)
OS: Fedora release 27 (Twenty Seven)
Linux version: 4.17.3-100.fc27.x86_64
gcc version: 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC)
SPDK: master (commit id: ee9db7dac0ee61a85f56dc8d8353b02205d0e6ad)

**Comment by pawelkax** [comment]
@ChenWang01 I cannot reproduce this error. I work on VM with Ubuntu 16.04. 
This kind of test is in patch https://review.gerrithub.io/#/c/416559/ where only one nvme is created and deleted. I  also cannot reproduce this error with this patch.

**Comment by ChenWang01** [comment]
Hi, pawel
@pawelkax
I have retried the issue with master(70ec42ad6a64eb1f0c0717d6ee72dfe7bd735f12), it's still exist.
The step 5 is necessary, step 4 and step 5 must use the same name to create the bdev, Then delete the nvme0n1, the issue will reproduce.

**Comment by pawelkax** [comment]
I get following error after step 5, and after killing spdk_tgt:
bdev.c:2809:spdk_bdev_init: *ERROR*: Bdev name:nvme0n1 already exists
bdev_nvme.c:1059:spdk_bdev_nvme_create: *ERROR*: Failed to find new NVMe controller
EAL: Cannot find unplugged device (0000:00:05.0)
bdev_nvme.c:1059:spdk_bdev_nvme_create: *ERROR*: Failed to find new NVMe controller
^C
=================================================================
==17735==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 312 byte(s) in 1 object(s) allocated from:
    #0 0x7f0d0aa1579a in __interceptor_calloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x9879a)
    #1 0x4915af in create_ctrlr /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:921
    #2 0x4920bc in attach_cb /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:983
    #3 0x4ecf6f in nvme_init_controllers /root/cli_pn_spdk/lib/nvme/nvme.c:452
    #4 0x4ed9f5 in spdk_nvme_probe_internal /root/cli_pn_spdk/lib/nvme/nvme.c:559
    #5 0x4edc42 in spdk_nvme_probe /root/cli_pn_spdk/lib/nvme/nvme.c:588
    #6 0x4929e0 in spdk_bdev_nvme_create /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:1051
    #7 0x498ec1 in spdk_rpc_construct_nvme_bdev /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme_rpc.c:136
    #8 0x717656 in spdk_jsonrpc_handler /root/cli_pn_spdk/lib/rpc/rpc.c:81
    #9 0x71ea5f in spdk_jsonrpc_server_handle_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:183
    #10 0x71990c in parse_single_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server.c:116
    #11 0x71a52d in spdk_jsonrpc_parse_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server.c:178
    #12 0x71f2a3 in spdk_jsonrpc_server_conn_recv /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:242
    #13 0x720c54 in spdk_jsonrpc_server_poll /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:386
    #14 0x7182c2 in spdk_rpc_accept /root/cli_pn_spdk/lib/rpc/rpc.c:199
    #15 0x6cfc9f in spdk_rpc_subsystem_poll /root/cli_pn_spdk/lib/event/rpc.c:51
    #16 0x6cd55d in _spdk_reactor_run /root/cli_pn_spdk/lib/event/reactor.c:541
    #17 0x6ce9a4 in spdk_reactors_start /root/cli_pn_spdk/lib/event/reactor.c:692
    #18 0x6c639b in spdk_app_start /root/cli_pn_spdk/lib/event/app.c:575
    #19 0x408140 in main /root/cli_pn_spdk/app/spdk_tgt/spdk_tgt.c:121
    #20 0x7f0d077a682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Direct leak of 6 byte(s) in 1 object(s) allocated from:
    #0 0x7f0d0a9df30f in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x6230f)
    #1 0x4919a0 in create_ctrlr /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:932
    #2 0x4920bc in attach_cb /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:983
    #3 0x4ecf6f in nvme_init_controllers /root/cli_pn_spdk/lib/nvme/nvme.c:452
    #4 0x4ed9f5 in spdk_nvme_probe_internal /root/cli_pn_spdk/lib/nvme/nvme.c:559
    #5 0x4edc42 in spdk_nvme_probe /root/cli_pn_spdk/lib/nvme/nvme.c:588
    #6 0x4929e0 in spdk_bdev_nvme_create /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme.c:1051
    #7 0x498ec1 in spdk_rpc_construct_nvme_bdev /root/cli_pn_spdk/lib/bdev/nvme/bdev_nvme_rpc.c:136
    #8 0x717656 in spdk_jsonrpc_handler /root/cli_pn_spdk/lib/rpc/rpc.c:81
    #9 0x71ea5f in spdk_jsonrpc_server_handle_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:183
    #10 0x71990c in parse_single_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server.c:116
    #11 0x71a52d in spdk_jsonrpc_parse_request /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server.c:178
    #12 0x71f2a3 in spdk_jsonrpc_server_conn_recv /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:242
    #13 0x720c54 in spdk_jsonrpc_server_poll /root/cli_pn_spdk/lib/jsonrpc/jsonrpc_server_tcp.c:386
    #14 0x7182c2 in spdk_rpc_accept /root/cli_pn_spdk/lib/rpc/rpc.c:199
    #15 0x6cfc9f in spdk_rpc_subsystem_poll /root/cli_pn_spdk/lib/event/rpc.c:51
    #16 0x6cd55d in _spdk_reactor_run /root/cli_pn_spdk/lib/event/reactor.c:541
    #17 0x6ce9a4 in spdk_reactors_start /root/cli_pn_spdk/lib/event/reactor.c:692
    #18 0x6c639b in spdk_app_start /root/cli_pn_spdk/lib/event/app.c:575
    #19 0x408140 in main /root/cli_pn_spdk/app/spdk_tgt/spdk_tgt.c:121
    #20 0x7f0d077a682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: 318 byte(s) leaked in 2 allocation(s).

**Comment by pawelkax** [comment]
@ChenWang01 @benlwalker Could you please look at my last comment and confirm it reproduces?

**Comment by ChenWang01** [comment]
@pawelkax 
You can try to do step 6 after step 5 without kill the nvmf_tgt if you  can see the same issue.
I think it's the same issue that caused by repeating to create the bdev with the same name.

**Comment by Maciej-Szwed** [comment]
I've got same result as pawelkax with emulated NVMes on VM. The original trace looks weird as it comes from blobstore which is not used at all in the reproduction steps.

**Comment by Maciej-Szwed** [comment]
I'm getting the same result with physical devices.

**Comment by jimharris** [comment]
Hi @Maciej-Szwed and @pawelkax 

Can we get an updated sequence of steps to reproduce this issue?  Original submitter had 6 steps and issue was related to deleting the bdev (in a way that we don't support any more).  But latest seems to indicate the failure is now in killing the spdk_tgt after step 5.

Maybe the original issue should be closed, and we should have a new issue for this latest sequence?

-Jim

**Comment by Maciej-Szwed** [comment]
@jimharris reproduction steps are the same. Only change is that you have to shut down app after those steps to get that error. I'll investigate this with new delete call and file new issue for the leaks we that we saw.

**Comment by Maciej-Szwed** [comment]
There is the same issue with the new delete command. I've created patch that fixes this:
https://review.gerrithub.io/#/c/spdk/spdk/+/420845

**Comment by ChenWang01** [comment]
@Maciej-Szwed 
I have retested this issue with your patch https://review.gerrithub.io/#/c/spdk/spdk/+/420845
The ASAN error can't reproduced, but the bdev can't be deleted after the step 4 and step 5.
I think it's still needs to  be solved.
```
/bdevs/nvme> delete nvme0n1
Got JSON-RPC error response
request:
{
  "params": {
    "name": "nvme0n1"
  },
  "jsonrpc": "2.0",
  "method": "delete_nvme_controller",
  "id": 1
}
response:
{
  "message": "No such device",
  "code": -32602
}
/bdevs/nvme> ls
o- nvme ................................................................................................................. [Bdevs: 1]
  o- nvme0n1 ................................................................................................ [Size=745.2G, Claimed]
```

**Comment by Maciej-Szwed** [comment]
Recently some patches that changing delete behavior has been merged. Could you try "delete nvme0" instead of "delete nvme0n1" ?

**Comment by ChenWang01** [comment]
@Maciej-Szwed 
i used the "delete nvme0" command and it can delete the nvme0n1 bdev, but we use the "ls" command to see the bdev name is "nvme0n1", there is something to be improved.
Another issue is:
We can't create bdev use `create name=nvme1 trtype=PCIe traddr=0000.04.00.0` after step 1 and step 2
```
1. create name=nvme0 trtype=PCIe traddr=0000.03.00.0
2. create name=nvme0 trtype=PCIe traddr=0000.04.00.0
3. create name=nvme1 trtype=PCIe traddr=0000.04.00.0
```
And we can create the bdev as following
```
1. create name=nvme0 trtype=PCIe traddr=0000.03.00.0
2. create name=nvme1 trtype=PCIe traddr=0000.04.00.0
```

**Comment by ChenWang01** [comment]
The ASAN error can't be reproduced after https://review.gerrithub.io/#/c/spdk/spdk/+/420845 this patch merged.
So i will close this issue.
